examples/seismic/source.py

```diff
@@ -0,0 +1,47 @@
+from devito.dimension import Dimension, time
+from devito.pointdata import PointData
+from devito.logger import error
+
+
+__all__ = ['PointSource', 'Receiver', 'Shot']
+
+
+class PointSource(PointData):
+    """Symbolic data object for a set of sparse points sources
+
+    :param name: Name of the symbol representing this source
+    :param coordinates: Point coordinates for this source
+    :param data: (Optional) Data values to initialise point data
+    :param ntime: (Optional) Number of timesteps for which to allocate data
+    :param npoint: (Optional) Number of sparse points represented by this source
+    :param dimension: :(Optional) class:`Dimension` object for
+                       representing the number of points in this source
+
+    Note, either the dimensions `ntime` and `npoint` or the fully
+    initialised `data` array need to be provided.
+    """
+
+    def __new__(cls, name, coordinates, data=None, ntime=None, **kwargs):
+        p_dim = kwargs.get('dimension', Dimension('p_%s' % name))
+        npoint = coordinates.shape[0]
+        if data is None:
+            if ntime is None:
+                error('Either data or ntime are required to'
+                      'initialise source/receiver objects')
+        else:
+            ntime = ntime or data.shape[0]
```

Comment:
Allowing the shape of the data to be different from the provided number of timesteps might be opening up to a range of potential mistakes by app developers.